### PR TITLE
[while_loop] data-dependent op in body_fn

### PR DESCRIPTION
**Pull request by ydwu4**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #141393
* #140106
* __->__ #142031
* #142162

The idea is the parent hop's fake tensor mode should ignore the newly allocated unbacked symints in subgraph because the bindings of unbacked symbols in the subgraph should already be done when we trace the subgraph. E.g. if there's an operator in subgraph that produces unbacked symints, the track_tensor_tree logic for that operator will take care of it.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

File: torch/_higher_order_ops/while_loop.py [hunk 1]
```diff
@@ -250,7 +248,41 @@ def while_loop_fake_tensor_mode(
...
+        # Case 2.2:
+        # if the unbacked symints are part of carried_inputs e.g.
+        #   nz = a.nonzero()
+        #   body_fn(it, nz):
+        #     return it+1. nz.sin() + 1,
+        # There's no new unbacked symints allocated in subgraph, so we're safe.
+        with mode.shape_env.ignore_fresh_unbacked_symbols():
+            # body_fn return output with the same pytree and tensor meta data as carried_inputs
+            # so we could just return the output after one iteration.
+            return body_fn(*carried_inputs, *additional_inputs)
```
Comment (zou3519) [inline torch/_higher_order_ops/while_loop.py:285]: discussed offline the metadata check should be in the fake kernel here

Comment (ydwu4) [reply torch/_higher_order_ops/while_loop.py]: Done! We cannot share the check (we could share the utils though) with dynamo as dynamo doesn't do the fake propagation of while_loop so we have to have two checks: one in fake tensor mode, the other in dynamo..

File: torch/_higher_order_ops/while_loop.py [hunk 2]
```diff
@@ -245,12 +245,64 @@ def _trace_while_loop(
     )
 
 
+def check_outputs_carry_consistency(
+    outs: List[torch.Tensor], crys: List[torch.Tensor]
```
Comment (zou3519) [inline torch/_higher_order_ops/while_loop.py:249]: nit: crys -> carries

File: torch/_higher_order_ops/while_loop.py [hunk 3]
```diff
@@ -245,12 +245,64 @@ def _trace_while_loop(
...
+    all_diffs_in_meta = []
+    for out, cry in zip(outs, crys):
+        if diff := diff_tensor_meta(
+            _extract_tensor_metadata(cry), _extract_tensor_metadata(out)
+        ):
+            all_diffs_in_meta.append(",".join(diff))
+    if all_diffs_in_meta:
+        diff_str = "\n".join(all_diffs_in_meta)
+        raise RuntimeError(
+            f"Expected carried_inputs and body outputs return tensors with same metadata but find:\n{diff_str}"
```
Comment (zou3519) [inline torch/_higher_order_ops/while_loop.py:260]: nit: "but find" -> "but found"

**Review by zou3519 — APPROVED** [review]

**Pull request closed without merging**